l1-contracts/contracts/chain-registrar/ChainRegistrar.sol

```diff
@@ -0,0 +1,180 @@
+import {IStateTransitionManager} from "../state-transition/IStateTransitionManager.sol";
+import {ETH_TOKEN_ADDRESS} from "../common/Config.sol";
+import {IERC20} from "@openzeppelin/contracts-v4/token/ERC20/IERC20.sol";
+import {Ownable2StepUpgradeable} from "@openzeppelin/contracts-upgradeable-v4/access/Ownable2StepUpgradeable.sol";
+import {IGetters} from "../state-transition/chain-interfaces/IGetters.sol";
+import {ReentrancyGuard} from "../common/ReentrancyGuard.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @dev ChainRegistrar serves as the main point for chain registration.
+contract ChainRegistrar is Ownable2StepUpgradeable, ReentrancyGuard {
+    /// Address that will be used for deploying l2 contracts
+    address public l2Deployer;
+    /// ZKsync Bridgehub
+    IBridgehub public bridgehub;
+
+    /// Chains that has been successfully deployed
+    mapping(bytes32 => bool) public deployedChains;
+    /// Proposal for chain registration
+    mapping(bytes32 => ChainConfig) public proposedChains;
+
+    error ProposalNotFound();
+    error BaseTokenTransferFailed();
+    error ChainIsAlreadyDeployed();
+    error ChainIsNotYetDeployed();
+    error BridgeIsNotRegistered();
+
+    /// @notice new chain is deployed
+    event NewChainDeployed(uint256 indexed chainId, address author, address diamondProxy, address chainAdmin);
+
+    /// @notice new chain is proposed to register
+    event NewChainRegistrationProposal(uint256 indexed chainId, address author, bytes32 key);
+
+    /// @notice Shared bridge is registered on l2
+    event SharedBridgeRegistered(uint256 indexed chainId, address l2Address);
+
+    /// @notice L2 Deployer has changed
+    event L2DeployerChanged(address newDeployer);
+
+    struct BaseToken {
+        address tokenAddress;
+        address tokenMultiplierSetter;
+        uint128 gasPriceMultiplierNominator;
+        uint128 gasPriceMultiplierDenominator;
+    }
+
+    struct ChainConfig {
+        /// @param Chain id of the new chain should be unique for this bridgehub
+        uint256 chainId;
+        /// @param Operator for making commit txs.
+        address blobOperator;
+        /// @param Operator for making Prove and Execute transactions
+        address operator;
+        /// @param Governor of the chain. Ownership of the chain will be transferred to this operator
+        address governor;
+        /// @param baseToken of the chain
+        BaseToken baseToken;
+        /// @param pubdataPricingMode How the users will charged for pubdata.
+        PubdataPricingMode pubdataPricingMode;
+    }
+
+    struct RegisteredChainConfig {
+        address pendingChainAdmin;
+        address chainAdmin;
+        address diamondProxy;
+        address l2BridgeAddress;
+    }
+
+    // @dev Initialize the contract
+    function initialize(address _bridgehub, address _l2Deployer, address _owner) public reentrancyGuardInitializer {
```

Comment:
It this contract is not on the proxy - then you need to move this logic to constructor. Otherwise, you would need to add `reentrancyGuardInitializer` on the implementation

Reply:
I think, it should be a proxy, that why I moved it to initialize

Reply:
I mean, you need to add constructor with `reentrancyGuardInitializer` modifier